src/query_error.c

```diff
@@ -34,61 +55,91 @@ const char *QueryError_Strerror(QueryErrorCode code) {
   return "Unknown status code";
 }
 
-void QueryError_SetError(QueryError *status, QueryErrorCode code, const char *err) {
-  if (status->code != QUERY_OK) {
+//---------------------------------------------------------------------------------------------
+
+/**
+ * Set the error code of the query. If `err` is present, then the error
+ * object must eventually be released using QueryError_Clear().
+ *
+ * Only has an effect if no error is already present
+ */
+
+void QueryError::SetError(QueryErrorCode c, const char *err) {
+  if (code != QUERY_OK) {
     return;
   }
-  RS_LOG_ASSERT(!status->detail, "detail of error is missing");
-  status->code = code;
+  RS_LOG_ASSERT(!detail, "detail of error is missing");
+  code = c;
 
   if (err) {
-    status->detail = rm_strdup(err);
+    detail = rm_strdup(err);
   } else {
-    status->detail = rm_strdup(QueryError_Strerror(code));
+    detail = rm_strdup(Strerror(code));
   }
```

Comment:
That the `strdup` is here because `detail` should be `Clear`-ed later. However, notice that `QueryError` holds a dynamically-allocated member but not a destructor, which makes it a very sick dog. We cannot just add an destructor with `rm_free` because someone is probably relying on the `detail` data to be around after the `QueryError` object is gone, and this looks like a problem we better avoid at this stage.